### PR TITLE
[mle] simplify the channel mask

### PR DESCRIPTION
**Pull request by zhanglongxia**

### CODE REVIEW

File: src/core/thread/mesh_forwarder.cpp [hunk 1]
```diff
@@ -1146,9 +1139,11 @@ void MeshForwarder::HandleSentFrame(Mac::Frame &aFrame, otError aError)
     }
 }
 
-void MeshForwarder::SetDiscoverParameters(uint32_t aScanChannels)
+void MeshForwarder::SetDiscoverParameters(const Mac::ChannelMask &aScanChannels)
 {
-    mScanChannels = (aScanChannels == 0) ? static_cast<uint32_t>(Mac::kScanChannelsAll) : aScanChannels;
+    uint32_t mask;
```
Comment (bukepo) [inline src/core/thread/mesh_forwarder.cpp]: add blank line below. see code style guide.

Comment (zhanglongxia) [reply src/core/thread/mesh_forwarder.cpp]: Changed.

File: src/core/thread/mesh_forwarder.hpp [hunk 2]
```diff
@@ -140,7 +141,7 @@ class MeshForwarder : public InstanceLocator
      * @param[in]  aScanChannels  A bit vector indicating which channels to scan.
```
Comment (abtink) [inline src/core/thread/mesh_forwarder.hpp]: nit: changing the comment about the paramter.
Also maybe adding a comment that if `aScanChannels` is empty, then all channels are used instead.

File: src/core/thread/mle.cpp [hunk 3]
```diff
@@ -2379,29 +2386,17 @@ otError Mle::SendAnnounce(uint8_t aChannel, bool aOrphanAnnounce, const Ip6::Add
...
 
-    VerifyOrExit(mAnnounceChannel <= OT_RADIO_CHANNEL_MAX, error = OT_ERROR_NOT_FOUND);
+    SuccessOrExit(error = GetActiveDatasetChannelMask(channelMask));
 
-    while (!channelMaskEntry->IsChannelSet(mAnnounceChannel))
+    if (channelMask.GetNextChannel(mAnnounceChannel) != OT_ERROR_NONE)
     {
-        mAnnounceChannel++;
-        VerifyOrExit(mAnnounceChannel <= OT_RADIO_CHANNEL_MAX, error = OT_ERROR_NOT_FOUND);
+        ExitNow(error = OT_ERROR_NOT_FOUND);
```
Comment (abtink) [inline src/core/thread/mle.cpp]: suggestion: How about simplifying this with `SuccessOrExit()`? (`GetNextChannel()` returns `OT_ERROR_NOT_FOUND` on failure). 
```
SuccessOrExit(error = channelMask.GetNextChannel(mAnnounceChannel));
```
in place of 
```
if (channelMask.GetNextChannel(mAnnounceChannel) != OT_ERROR_NONE)
{ 
    ExitNow(error = OT_ERROR_NOT_FOUND);
}
```

File: src/core/thread/mesh_forwarder.hpp (hunk 2 above)
Comment (zhanglongxia) [reply src/core/thread/mesh_forwarder.hpp]: Changed.

File: src/core/thread/mle.cpp (hunk 3 above)
Comment (zhanglongxia) [reply src/core/thread/mle.cpp]: Changed.

File: src/core/thread/mle.cpp [hunk 4]
```diff
@@ -1744,6 +1727,30 @@ bool Mle::PrepareAnnounceState(void)
     return shouldAnnounce;
 }
 
+otError Mle::GetActiveDatasetChannelMask(Mac::ChannelMask &aChannelMask)
```
Comment (jwhui) [inline src/core/thread/mle.cpp]: There seems to be another `GetActiveDatasetChannelMask()` in [src/core/thread/announce_sender.cpp](https://github.com/openthread/openthread/blob/c06179938c8d5deb59822f28482304e2589baf69/src/core/thread/announce_sender.cpp#L125). Can we combine these into a single implementation?

Comment (abtink) [reply src/core/thread/mle.cpp]: How about adding this method to `ActiveDataset` or `DatasetManager` class?

Comment (zhanglongxia) [reply src/core/thread/mle.cpp]: Great suggestion. I have added a method to `DatasetManager ` class.

**Pull request merged**